src/dynout_parser.cc

```diff
@@ -0,0 +1,135 @@
+/* Generated by re2c 1.3 */
+// Copyright 2011 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dynout_parser.h"
+
+#include "debug_flags.h"
+#include "metrics.h"
+
+#include <errno.h>
+
+using namespace std;
+
+namespace {
+
+size_t find_newline(const string& str, size_t pos) {
+    for (size_t i = pos; i < str.size(); i++) {
+      if (str[i] == '\r' || str[i] == '\n') {
+        return i;
+      }
+    }
+    return std::string::npos;
+}
+
+}
+bool DynoutParser::Parse(State* state, DiskInterface* disk_interface,
+            Edge* edge, const std::string& path, 
+            std::vector<Node*>* nodes, int* outputs_count,
+            std::string* err) {
+METRIC_RECORD("dynout load");
+  // Read depfile content.  Treat a missing depfile as empty.
+  string content;
+  switch (disk_interface->ReadFile(path, &content, err)) {
+  case DiskInterface::Okay:
+    break;
+  case DiskInterface::NotFound:
+    if (err != NULL) {
+      err->clear();
+    }
+    break;
+  case DiskInterface::OtherError:
+    if (err != NULL) {
+      *err = "loading '" + path + "': " + *err;
+    }
+    return false;
+  }
+  if (content.empty()) {
+    return true;
+  }
+
+  vector<string> output_paths;
+  int start_size = nodes->size();
+
+  size_t pos = 0;
+  size_t next = find_newline(content, pos);
+  while (next != std::string::npos) {
+    if (next > pos) {
+      output_paths.push_back(content.substr(pos, next - pos));
+    }
+    pos = next + 1;
```

Comment:
Mmm. this code handles \r\n by looping twice which is sub-optimal due to the extra function calls. Also, you should be able to write a single function that parses the content and returns a vector of StringPiece instead to avoid needless allocations and string copies, just like the depfile parser does. E.g.

```
/// Split `content` into a series of individual text lines, without trailing newlines.
/// Note that this also removes empty lines from the result.
std::vector<StringPiece> SplitTextLines(StringPiece content) {
  std::vector<StringPiece> result;
  const char* input = content.str_;
  const char* start = input;
  const char* limit = input + content.str_;
  while (start < limit) {
    const char* end = start;
    // Note: \r\n will be treated as a line end + an empty line,
    // the latter will be ignored. This simplifies the logic in this
    // loop to reduce its code and speed it up.
    while (end < limit && (*end != '\n' || *end != '\r'))
      ++end;
    if (end > start)
      result.append(StringPiece(start, end));
    start = end + 1;
  }
  return result;
}
```

Reply:
Good call, `DynoutParser` now returns a vector of `StringPiece`.